### DISCUSSION TITLE
App state - Public multi-step form

### DISCUSSION THREAD

**Discussion by eyduh** [Budibase/budibase#7256] [Help]
### Hi
I've set up a public multistep form for submissions to a film festival and been experimenting with using app state to save the progress in the users browser cache, but I can't seem to get it to work properly. 

Do I need to add app state logic to every field of the form or is there a way to save all the fields of each form step when using the navigation buttons up top or down below?

At the moment I've been trying to use "On change" triggers on individual fields but seeing as I haven't managed to get it to work I'm looking for another route. Besides the not working part, It would be quite cumbersome to add it to every field of the entire form.

Had a look at the docs page and through the discussions but couldn't find a public form use case.

Cheers!

**Comment by andz-bb** [comment]
Hey @eyduh

I've been playing around with this today and have had some success with implementing something similar to what you are talking about. I will try my best to explain how I have set it up.

Essentially I have created an additional table with columns for all of the fields in my form to cache the user submission data (with identical column names/types to the original schema). When I complete the first step of my form a row is created in this new Form table and I then save the _id for that record to my app state (persistent state i.e. browser Local Storage). Each time I click the next button to move forward in the form, this row is updated with the data I have just entered. I also have another variable in my app state for the form step, so each time I move steps I also update that value. This means that if I close the window and come back, I could have a button that conditionally renders based on whether I have a form ID stored in the app state that has actions which will jump me back to the last page of the form I was on (using data provider/repeater for the form table filtered down using the app state form ID).

On the final submit at the end of the form I then use Save Row actions to save all of the form data to the original tables and remove the working data/app state. If the columns match up between your working form table and the real data tables it should work well. Honestly my setup for this was pretty complex so I might have missed something when typing this out. If you have any questions off the back of this let me know and I'll try and go into more detail about what I have done.

**Reply by eyduh** [reply]
Hi,

I think I get it and this was the way I suspected I would need to set it up. 

I think I will set up another database with a table that mirrors this one and try it out. At the moment my current setup will be enough but will be returning to this in the near future, there's another form of similar use case/structure that needs to be online by December so I have some time to play around c:
